core/src/main/scala/org/apache/spark/ui/scope/RDDOperationGraph.scala

```diff
@@ -119,18 +119,38 @@ private[ui] object RDDOperationGraph extends Logging {
       { if (stage.attemptId == 0) "" else s" (attempt ${stage.attemptId})" }
     val rootCluster = new RDDOperationCluster(stageClusterId, stageClusterName)
 
+    var rootNodeCount = 0
+    val addRDDIds = new mutable.HashSet[Int]()
+    val dropRDDIds = new mutable.HashSet[Int]()
+
     // Find nodes, edges, and operation scopes that belong to this stage
-    stage.rddInfos.foreach { rdd =>
-      edges ++= rdd.parentIds.map { parentId => RDDOperationEdge(parentId, rdd.id) }
+    stage.rddInfos.sortBy(_.id).foreach { rdd =>
+      var isAllowed = true
+      val parentIds = rdd.parentIds
+      if (parentIds.isEmpty) {
+        isAllowed = rootNodeCount < retainedNodes
+        rootNodeCount += 1
+      } else {
+        isAllowed = parentIds.exists(id => addRDDIds.contains(id) || !dropRDDIds.contains(id))
+      }
+
+      if (isAllowed) {
+        addRDDIds += rdd.id
+        edges ++= parentIds.filter(id => !dropRDDIds.contains(id))
+          .map { parentId => RDDOperationEdge(parentId, rdd.id) }
```

Comment:
Also consider simplifying blocks like this to `.map(RDDOperationEdge(_, rdd.id))`